### PR TITLE
Added storage_path and fixed artifact.files, thanks @alvarobartt!

### PR DESCRIPTION
**Pull request by vanpelt**
Fixes WB-9899

Description
-----------
This adds a new `storage_path` accessor to `artifact.files`.  It also ended up being a bit of a rabbit hole.  New additions include:

1. Moved `artifact._files(...)` to `artifact.files` and made it actually work by using the right artifact accessors.
2. Added a new `api.client.server_info` method to ask the server what versions of the sdk it supports.
3. Made the `ArtifactFiles` query only ask for `storagePath` if we're sure it has it.  There's probably a better way to do this but it works.
4. Made the `File` class use our magic `Attrs` class to simplify it's implementation.  Made me want to figure out how to get those methods to be automatically documented 🤔 

Syncing mockserver with:
- https://github.com/wandb/yea-wandb/pull/85
- https://github.com/wandb/yea/pull/33

Testing
-------
Added new mocks and tested both the new server and legacy server case.

### CODE REVIEW

File: wandb/apis/public.py [hunk 1]
```diff
@@ -217,6 +233,17 @@ def execute(self, *args, **kwargs):
                 )
             raise
 
+    @property
+    def server_info(self):
+        if self._server_info is None:
+            self._server_info = self.execute(self.INFO_QUERY).get("serverInfo")
+        return self._server_info
+
+    def version_supported(self, min_version):
```
Comment (annirudh) [inline wandb/apis/public.py:242]: This is a really nice helper -- I think I've scattered in a bunch of ad-hoc checks for the other functions here.

**Review by annirudh — APPROVED** [review]
LGTM! Thank you so much for adding this.

**Comment by venky-wandb** [comment]
Awesome!! Thanks for adding this!! 💯 🥇

**Review by venky-wandb — APPROVED** [review]

**Review by alvarobartt — CHANGES_REQUESTED** [review]
Great work so far! 🎉 I was also planning also to fill a PR, as I'm working in [wandbfsspec](https://github.com/alvarobartt/wandbfsspec) and wanted to add not just the support for the File System but also to support the Artifact Store.

I checked your code and the following lines are not working:

https://github.com/wandb/wandb/blob/8dd418ba0a30da76f134743d53e7cd5b3938bfdb/wandb/apis/public.py#L4940-L4941

So I'd suggest either creating the `path` attribute and adding it to the print statement as:

```python
     return "<ArtifactFiles {} ({})>".format("/".join(self.artifact.path), len(self)) 
```

Or use the `name` instead:

```python
     return "<ArtifactFiles {} ({})>".format("/".join(self.artifact.name), len(self)) 
```

Edit: that also fails as the property `def length` is not implemented, so we'll need to create it too, I'll submit a PR over this branch in order to solve that #:hugs:

**Comment by alvarobartt** [comment]
@vanpelt I've already created a PR over this branch so as to fix `__repr__` over `ArtifactFiles`, hope it's useful! 🎉

**Pull request merged**